azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebApp.java

```diff
@@ -0,0 +1,302 @@
+    }
+
+    @Override
+    public void deploy(File file) {
+        deploy(AppServiceUtils.getDeployTypeByFileExtension(file), file);
+    }
+
+    public void deploy(DeployType deployType, File target) {
+        getWebAppClient().deploy(com.azure.resourcemanager.appservice.models.DeployType.fromString(deployType.getValue()), target);
+    }
+
+    @Override
+    public boolean exists() {
+        return getWebAppClient(true) != null;
+    }
+
+    @Override
+    public String hostName() {
+        return getWebAppClient().defaultHostname();
+    }
+
+    @Override
+    public String state() {
+        return getWebAppClient().state();
+    }
+
+    @Override
+    public PublishingProfile getPublishingProfile() {
+        final com.azure.resourcemanager.appservice.models.PublishingProfile publishingProfile = getWebAppClient().getPublishingProfile();
+        return PublishingProfile.createFromServiceModel(publishingProfile);
+    }
+
+    @Override
+    public Runtime getRuntime() {
+        return AppServiceUtils.getRuntimeFromWebApp(getWebAppClient());
+    }
+
+    @Override
+    public WebAppUpdater update() {
+        return new WebAppUpdater();
+    }
+
+    @Override
+    public IWebAppDeploymentSlot deploymentSlot(String slotName) {
+        final WebAppDeploymentSlotEntity slotEntity = WebAppDeploymentSlotEntity.builder().name(slotName)
+                .resourceGroup(getWebAppClient().resourceGroupName())
+                .webappName(getWebAppClient().name()).build();
+        return new WebAppDeploymentSlot(slotEntity, azureAppService);
+    }
+
+    @Override
+    public List<IWebAppDeploymentSlot> deploymentSlots() {
+        return getWebAppClient().deploymentSlots().list().stream()
+                .map(slot -> new WebAppDeploymentSlot(WebAppDeploymentSlotEntity.builder().id(slot.id()).build(), azureAppService))
+                .collect(Collectors.toList());
+    }
+
+    private com.azure.resourcemanager.appservice.models.WebApp getWebAppClient() {
+        return getWebAppClient(false);
+    }
+
+    synchronized com.azure.resourcemanager.appservice.models.WebApp getWebAppClient(boolean force) {
+        if (webAppClient == null || force) {
+            try {
+                webAppClient = StringUtils.isNotEmpty(entity.getId()) ?
+                        azureClient.webApps().getById(entity.getId()) :
+                        azureClient.webApps().getByResourceGroup(entity.getResourceGroup(), entity.getName());
+                entity = AppServiceUtils.getWebAppEntity(webAppClient);
+            } catch (RuntimeException e) {
+                // SDK will throw exception when resource not founded
+                webAppClient = null;
+            }
+        }
+        return webAppClient;
+    }
+
+    @Override
+    public String id() {
+        return getWebAppClient().id();
+    }
+
+    @Override
+    public String name() {
+        return getWebAppClient().name();
+    }
```

Comment:
is it necessary for us ? how about let entity() responsible for these basic properties ?

Reply:
As most usage of entity is to get the name and id, so offer id and name in manager interface directly